### PR TITLE
Fix false positives of IS_*() macros for 8-bit ASCII characters

### LINKED ISSUE #5778: Support 8-bit characters in conf_def.h
This issue was introduced by the removal of a contribution because we could not locate the original author to gain approval for the licence change.

After the code removal the various IS_\*() macros in the auto-generated file conf_def.h may incorrectly return true if the supplied character has its most significant bit set. The IS_\*() macros should be able to correctly handle 8-bit characters. Note that UTF-8 support is not a requirement.

### PR DESCRIPTION
**Pull request by mspncp**
Fixes #5778, #5840

The various `IS_*()` macros did not work correctly for 8-bit ASCII characters with the high bit set, because the `CVT(a)` preprocessor macro and'ed the given ASCII value with 0x7F, effectively folding the high value range 128-255 over the low value range 0-127. As a consequence, some of the `IS_*()` erroneously returned TRUE.

This commit fixes the issue by mapping CVT(a) to 127 for all values a >= 127. This works because the lookup tables CONF_type_default and CONF_type_win32 have all bits cleared at entry 127, whence IS_*(a) returns FALSE for all values outside the range 0-127.

The IS_*() macros were also changed to return TRUE or FALSE (1 or 0) instead of a nonzero or zero value.

Note that with this change, the macro `IS_*(c,a)` evaluates the `a` argument twice, unless CHARSET_EBCDIC is defined. This prohibits the use of arguments with side effects like `IS_*(c, *p++)`.

### CODE REVIEW

File: crypto/conf/conf_def.h [hunk 1]
```diff
@@ -28,21 +28,26 @@
...
-#define IS_NUMBER(c,a)      (KEYTYPES(c)[CVT(a)] & CONF_NUMBER)
-#define IS_WS(c,a)          (KEYTYPES(c)[CVT(a)] & CONF_WS)
-#define IS_ALNUM(c,a)       (KEYTYPES(c)[CVT(a)] & CONF_ALNUM)
-#define IS_ALNUM_PUNCT(c,a) (KEYTYPES(c)[CVT(a)] & CONF_ALNUM_PUNCT)
-#define IS_QUOTE(c,a)       (KEYTYPES(c)[CVT(a)] & CONF_QUOTE)
-#define IS_DQUOTE(c,a)      (KEYTYPES(c)[CVT(a)] & CONF_DQUOTE)
+/*
+ * Attention: because the macro argument 'a' is evaluated twice in CVT(a),
+ * it is not allowed pass 'a' arguments with side effects to IS_*(c,a)
+ * like for example IS_*(c, *p++).
```
Comment (levitte) [inline crypto/conf/conf_def.h]: I'm sorry, but I think that introducing this kind of fragility is the wrong way to go.

Comment (richsalz) [reply crypto/conf/conf_def.h]: But note that this header file is used/included in only one internal source file.

Comment (levitte) [reply crypto/conf/conf_def.h]: So we only shoot our own kneecaps off...  still.

**Comment by paulidale** [comment]
How much of this overlaps with`crypto/ctype.h` ?  Could this functionality be wrapped up by the former?

File: crypto/conf/conf_def.h (hunk 1 above)
Comment (richsalz) [reply crypto/conf/conf_def.h]: We're stuck with moving forward in pieces, I guess, because we had to remove the previous code.

**Comment by richsalz** [comment]
There's an open issue on this and ctype.  I forget the number.  There are some complications (one assumes ascii only?) but it could be worked around.

**Comment by richsalz** [comment]
I suppose another option is to make the table be 256 entries and fill the top 128 with 0

**Comment by paulidale** [comment]
#5840 ?  `crypto/ctype.h` knows about EBCDIC.  It does limit itself to 7 bit characters once EBCDIC is dealt with.

**Comment by richsalz** [comment]
No, #5672 (which I created!!) which refers to a comment in #5669.

**Comment by paulidale** [comment]
Glibc uses a table 192 long which goes from -128 to 255 to deal with signed chars quickly.
I'm not sure this is necessary here.  Checking the eighth bit is sufficient to catch both cases quickly.

**Comment by mspncp** [comment]
> I suppose another option is to make the table be 256 entries and fill the top 128 with 0

This sounds like a good idea. But I'll won't be able to do it before this evening, because I'm traveliing today.

**Comment by levitte** [comment]
I just had a look.  It's not *quite* as simple as one might think...

**Comment by mspncp** [comment]
> I just had a look.  It's not *quite* as simple as one might think...

Richard, I am aware that the macros IS_ALNUM*() do not work correctly for utf-8 characters (in particular german umlauts ;-) ), But according to @mattcaswell, it was not a requirement that these macros support utf-8 encodings in general (https://github.com/openssl/openssl/issues/5778#issue-309414218). So maybe it is only the exaggerated title of the commit message that disturbs you?

> Make the IS_*() macros work correctly for 8-bit ASCII characters

What I fixed was only the false positives caused by the fact that the range 0-255 was mapped two-to-one onto 0-127. So would you accept this pull request if I change the title of the commit message to something like the following?

> Fix false positives of IS_*() macros for 8-bit ASCII characters

**Comment by mspncp** [comment]
Or would you prefer if I fix the IS_*() macros to work correctly for all single byte utf-8 characters?

**Comment by levitte** [comment]
Sorry, my comment was about trying to use our ctype.h instead. *That* is a harder endeavor than one might think at first.

This PR looks good to me at this point. However, I cannot be the one to approve it in this particular case. It has to be someone else.

**Comment by mspncp** [comment]
This will be the final commit message:

```
Fix false positives of IS_*() macros for 8-bit ASCII characters

Fixes #5778, #5840

The various IS_*() macros did not work correctly for 8-bit ASCII
characters with the high bit set, because the CVT(a) preprocessor
macro and'ed the given ASCII value with 0x7F, effectively folding
the high value range 128-255 over the low value range 0-127.
As a consequence, some of the IS_*() erroneously returned TRUE.

This commit fixes the issue by enlarging the arrays CONF_type_default
and CONF_type_win32 to 256 entries, filling the upper 128 entries
with zeroes. In other words, IS_*(c, a) returns FALSE for all
arguments 128 <= a < 256.

The IS_*() macros were also changed to return TRUE or FALSE (1 or 0)
instead of a nonzero or zero value.
```

ping @mattcaswell

**Comment by mattcaswell** [comment]
> ping @mattcaswell

Thanks for all your work on this!! As I removed the original code I think someone from the OMC who was not involved with that removal should approve this. Maybe one of @vdukhovni, @iamamoose or @ekasper could do this?

**Comment by mattcaswell** [comment]
Or @kroeckx, or @t-j-h!

**Comment by mspncp** [comment]
I will also wait for @petrovr to confirm that this pull request indeed fixes #5840 .

**Comment by petrovr** [comment]
I think that mask in CVT is required, i.e. CVT(a) ((unsigned char)(a) & 0xFF) - on some platforms char is not 8-bit.
I'm not sure why TRUE or FALSE is preferred instead non-zero and zero. 

Everything that restrict those IS_* macros to work only on 7-bit ASCII characters is solution.
In particular #5844 should resolve #5840 .

**Comment by mspncp** [comment]
> I think that mask in CVT is required, i.e. CVT(a) ((unsigned char)(a) & 0xFF) - on some platforms char is not 8-bit.

@petrovr, thank you for pointing out a possible problem CHAR_BIT > 8. Thinking about it for a while, I noticed that your suggestion to use the 0xFF mask reintroduces exactly the same problem which this pull request tries to fix: If you take a 'char' with more than 8 bits and mask it with 0xFF, then you might end up with a 7 bit ASCII char and false positives for IS_*() again.

The only solution is a correct range check, which was not possible using preprocessor macros, without evaluating the macro argument multiple times. So solve this dilemma, I moved a lot of the implementation into a static function.

For better comparison, I created a separate pull request, see #5903.

**Comment by mspncp** [comment]
Closed in favour of #5903.

**Comment by petrovr** [comment]
10x Matthias.
I did some test with static inline function but due to Easter holidays I was not able to share results.
So  #5903 was the right solution.

**Pull request closed without merging**